File: scala-backend/src/main/scala/repository/ProfileRepository.scala
package repository

import scala.util.chaining._

import domain._
import slick.jdbc.MySQLProfile.api._
import scala.concurrent.{Future, ExecutionContext}

class ProfileRepository(db: Database)(implicit ec: ExecutionContext) {

  private class ParentsTable(tag: Tag) extends Table[ParentProfile](tag, "parents") {
    def id = column[Long]("id", O.PrimaryKey, O.AutoInc)
    def name = column[String]("name")
    def child = column[String]("child")
    def * = (id, name, child) <> (ParentProfile.tupled, ParentProfile.unapply)
  }

  private class PaymentMethodsTable(tag: Tag) extends Table[PaymentMethod](tag, "payment_methods") {
    def id = column[Long]("id", O.PrimaryKey, O.AutoInc)
    def parentId = column[Long]("parent_id")
    def method = column[String]("method")
    def isActive = column[Boolean]("is_active")
    def * = (id, parentId, method, isActive) <> (PaymentMethod.tupled, PaymentMethod.unapply)
  }

  private class InvoicesTable(tag: Tag) extends Table[Invoice](tag, "invoices") {
    def id = column[Long]("id", O.PrimaryKey, O.AutoInc)
    def parentId = column[Long]("parent_id")
    def amount = column[Double]("amount")
    def date = column[String]("date")
    def * = (id, parentId, amount, date) <> (Invoice.tupled, Invoice.unapply)
  }

  private val parents = TableQuery[ParentsTable]
  private val paymentMethods = TableQuery[PaymentMethodsTable]
  private val invoices = TableQuery[InvoicesTable]

  def createPaymentMethod(paymentMethod: PaymentMethod): Future[Either[String, PaymentMethod]] =
    db.run((paymentMethods returning paymentMethods.map(_.id)
        into ((method, id) => method.copy(id = id))) += paymentMethod)
      .map(Right(_))
      .recover { case ex => Left(s"Failed to add payment method: ${ex.getMessage}") }

  def retrievePaymentMethods(parentId: Long): Future[Seq[PaymentMethod]] =
    db.run(paymentMethods.filter(_.parentId === parentId).result)

  def retrieveInvoices(parentId: Long): Future[Seq[Invoice]] =
    db.run(invoices.filter(_.parentId === parentId).result)

  def retrieveParentProfiles(parentId: Long): Future[Seq[ParentProfile]] =
    db.run(parents.filter(_.id === parentId).result)

  def updatePaymentMethods(updatedPaymentMethods: Seq[PaymentMethod]): Future[Seq[Int]] =
    updatedPaymentMethods
      .map(paymentMethod => paymentMethods.filter(_.id === paymentMethod.id).update(paymentMethod))
      .pipe(DBIO.sequence(_).transactionally)
      .pipe(db.run)

  def deletePaymentMethod(methodId: Long): Future[Either[String, Int]] =
    db.run(paymentMethods.filter(pm => pm.id === methodId).delete)
      .map {
        case rowsDeleted if rowsDeleted > 0 => Right(rowsDeleted)
        case _ => Left(s"Payment method with id $methodId not deleted, because it wasn't there")
      }
      .recover { case ex => Left(s"Failed to delete payment method: ${ex.getMessage}") }
}

File: nodejs-backend/src/repository/profileRepository.ts
import { db, query } from '../db/database';
import mysql from 'mysql2/promise';
import { Invoice, ParentProfile, PaymentMethod } from '../parentProfileBackend';

export class ProfileRepository {
    async createPaymentMethod(
        paymentMethod: PaymentMethod
    ): Promise<PaymentMethod> {
        const sql =
            'INSERT INTO payment_methods (parent_id, method, is_active) VALUES (?, ?, ?)';
        const [result] = await db.execute<mysql.ResultSetHeader>(sql, [
            paymentMethod.parentId,
            paymentMethod.method,
            paymentMethod.isActive,
        ]);
        const insertId = result.insertId;
        return { ...paymentMethod, id: insertId };
    }

    async retrievePaymentMethods(parentId: number): Promise<PaymentMethod[]> {
        const sql = 'SELECT * FROM payment_methods WHERE parent_id = ?';
        const results = await query(sql, [parentId]);
        return results.map((r) => ({
            id: r.id,
            parentId: r.parent_id,
            method: r.method,
            isActive: r.is_active,
            createdAt: r.created_at.toLocaleString(),
        }));
    }

    async retrieveInvoices(parentId: number): Promise<Invoice[]> {
        const sql = 'SELECT * FROM invoices WHERE parent_id = ?';
        const results = await query(sql, [parentId]);
        return results.map((r) => ({
            id: r.id,
            parentId: r.parent_id,
            amount: r.amount,
            date: r.date.toLocaleString(),
        }));
    }

    async retrieveParentProfiles(parentId: number): Promise<ParentProfile[]> {
        const sql = 'SELECT * FROM parents WHERE id = ?';
        const results = await query(sql, [parentId]);
        return results.map((r) => ({
            id: r.id,
            name: r.name,
            child: r.child,
        }));
    }

    async updatePaymentMethods(
        updatedPaymentMethods: PaymentMethod[]
    ): Promise<number[]> {
        const updatePromises = updatedPaymentMethods.map((paymentMethod) => {
            const sql =
                'UPDATE payment_methods SET parent_id = ?, method = ?, is_active = ? WHERE id = ?';
            return db.execute<mysql.ResultSetHeader>(sql, [
                paymentMethod.parentId,
                paymentMethod.method,
                paymentMethod.isActive,
                paymentMethod.id,
            ]);
        });
        const results = await Promise.all(updatePromises);
        return results.map(([result]) => result.affectedRows);
    }

    async deletePaymentMethod(methodId: number): Promise<boolean> {
        const sql = 'DELETE FROM payment_methods WHERE id = ?';
        const [result] = await db.execute<mysql.ResultSetHeader>(sql, [
            methodId,
        ]);
        return result.affectedRows > 0;
    }
}
